### ISSUE TITLE
API simplification

### ISSUE THREAD

**Issue by sehz** [infinyon/fluvio#365]
For first OSS release, we should make  following updates to make it simpler with minimum changes.

For consumer.  add `RecordIterator`, similar to iterator on hash map.  It will return tuple of metadata and records.  So example  code in the consumer can be look like this:

```
async fn consume() -> Result<(), FluvioError> {
    let consumer = fluvio::consumer(TOPIC, 0).await?;
    let mut stream = consumer.stream(Offset::beginning()).await?;

    while let Some((_,record)?) = stream.next().await {
      
          let string = String::from_utf8(record.as_byte()).unwrap();
          println!("Got record: {}", string);
          if string == "Done!" {
                  return Ok(());
          }
          
    }

    Ok(())
}

**Comment by sehz** [comment]
Updated to add try operator in the while block since `next()` should return Result.   This should make it compatible with `try_next()` in the `StreamExt`

**Comment by nicholastmosher** [comment]
I'm not seeing a `try_next` method in the `StreamExt` trait here: https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html

I think I know what method you're talking about, but does it belong to a different trait or library? Can you find it and post it here?

**Comment by nicholastmosher** [comment]
Can we also start a discussion about what steps are needed in order to create a proper `Stream`-implementation? I'm not familiar enough with the inner workings of the consumer to know what the fetching process looks like.

**Comment by sehz** [comment]
Sorry discard comment about about `try_next`.   It is deprecated so shouldn't have to worry about it.  Important point is that `next` should return `Option<Result<...`

For stream implementation,  challenges is map futures -> stream which means you need to hold some state in order to map to stream.  Some example: https://github.com/infinyon/k8-api/blob/master/src/k8-client/src/list_stream.rs

If this is too complex,  we can leave as futures and address stream conversion in next release

**Comment by sehz** [comment]
From looking at the code, simplest implementation is to add `Stream` trait in: [AsyncResponse](https://github.com/infinyon/fluvio-socket/blob/master/crates/socket/src/multiplexing.rs)

**Comment by sehz** [comment]
Or you can implement something like: [BufferedStream](https://docs.rs/futures-util/0.3.6/src/futures_util/stream/stream/buffered.rs.html#14-23) over batch stream

**Comment by sehz** [comment]
Or use https://docs.rs/futures/0.3.6/futures/stream/trait.StreamExt.html#method.flat_map

**Comment by nicholastmosher** [comment]
Working on the first step for this here: https://github.com/infinyon/fluvio-socket/pull/4

Need to figure out how to make a stream of records rather than a stream of responses. Flat_map looks promising, but the implementing code might not know enough about R to break apart the response

**Comment by nicholastmosher** [comment]
On the `DefaultRecord` type, we usually have a flow that looks like this:

```rust
if let Some(bytes) = record.value.inner_value() { ... }
```

In what case would the `inner_value` ever be `None`? Would we be able to remove the `Option` here and just have the bytes always available?

**Comment by sehz** [comment]
I think was to make it compatible with both file slice and bytes.    Changes requires too much surgery for now

**Comment by nicholastmosher** [comment]
This will be closed once #372 is merged

**Comment by sehz** [comment]
done
